P12.py

```diff
@@ -0,0 +1,64 @@
+import os
+import hashlib
+import threading
+import math
+
+class FilesInDirectory:
+    def __init__ (self, dir:str):
+        self.dir = dir
+        self.duplicates = {}
+        self.duplicates_pos = []
+        self.files_list = []
+
+        self.__files_from_dir()
+
+    def __files_from_dir (self):
+        for dir in os.walk(self.dir):
+            for i in dir[2]:
+                self.files_list.append(dir[0]+"/"+i)
+
+    def get_duplicates (self, n):
+        files_shift = math.ceil(len(self.files_list)/n)
+        files_start = -files_shift
+        threads = []
+
+        while True:
+            files_start += files_shift
+            files_end = files_start+files_shift
+            if files_end > len(self.files_list): 
+                files_end = len(self.files_list)
+            t = threading.Thread(target=self.__get_duplicates_proccess, args=[files_start, files_end, 65536])
+            t.start()
+            threads.append(t)
+            if files_end == len(self.files_list): break
+
+        for t in threads:
+            t.join()
+
+        res = []
+        for i in self.duplicates_pos:
+            res.append(self.duplicates[i])
+
+        return res
+
+    def __get_duplicates_proccess (self, files_start, files_end, chunk_size):
+        for i in range(files_start, files_end):
+            hasher = hashlib.md5()
+            with open(self.files_list[i], 'rb') as file:
+                buf = file.read(chunk_size)
+            while len(buf) > 0:
+                hasher.update(buf)
+                buf = file.read(chunk_size)
+            file.close()

```

Comment:
if you use context manager, you don't need to close a file, it's done in __exit__ method of a context manager. Please read the context manager documentation carefully

Reply:
Ohh, right, I knew it. I just forgot to remove it